---
title: "R_Assignment_WAB"
output: html_document
---

R Assignment


--- installs packages from tidyverse
--- reads fang genotypes file into data frame
--- reads snp positions into data frame
--- selects SNP_ID, Chromosome, and Position columns to new data frame
--- filters genotypes to extract maize genotype data to new data frame, str_detect detects presence of match for desired specimen groups, filters those
--- filters genotypes to extract teosinte genotype data to new data frame, str_detect detects presence of match for desired specimen groups, filters those
--- removes header info from maize snps, transposes, defines as data frame, turns row names into columns, renames as SNP_ID to facilitate joining to snps
--- removes header info from teosinte snps, transposes, defines as data frame, turns row names into columns, renames as SNP_ID to facilitate joining to snps
--- merges SNP info with maize genotype data
--- merges SNP info with teosinte genotype data
--- removes SNPs with multiple or unknown positions from maize SNPs
--- removes SNPs with multiple or unknown positions from teosinte SNPs
--- sorts maize SNPs by chromosome, then position ascending, numerically
--- sorts teosinte SNPs by chromosome, then position ascending, numerically
--- sorts maize SNPs by chromosome, then position descending, numerically to separate df 
--- sorts teosinte SNPs by chromosome, then position descending, numerically to separate df 
--- replaces ?/? with -/- in descending maize SNPs file
--- replaces ?/? with -/- in descending teosinte SNPs file



TODO
  maizesplit <- split(maize_snps, maize_snps$Chromosome)
  maizesplit_desc <- split(maize_snps_desc, maize_snps_desc$Chromosome)

for(i in names(maizesplit)) {
    assign(paste0("maize_", i), maizesplit[[i]])
}
  above functions split maize into separate data frames for each chromosome - run after sorting
  
  for other 10 files, will need to desc_maize_snps<-desc(maize_snps, as.numeric(Chromosome), as.numeric(Position)) and then split into desc_maize_i files - also need to replace ?/? with -/-
  
  working order 
  - split into files and print to .txt for each maize/teosinte ascending/descending
 - visualize
 - data inspection
  
for visualization, have one file with teosinte/maize snps, including group info? filter as before for maize/teo groups into one file, transpose (-(Sample_ID:JG_OTU)), merge, filter unknown/multiple - can visualize by group - make sure group column named as group if making one large file

  
```{r}
rm(list = ls())
library(tidyverse)
fang_genos <- read_tsv("Data/fang_et_al_genotypes.txt")
snp_pos <- read_tsv("Data/snp_position.txt")
snps <- select(snp_pos, SNP_ID, Chromosome, Position)
maize_genos <- fang_genos %>% 
    filter(str_detect(Group, "ZMMIL|ZMMLR|ZMMMR"))
teosinte_genos <- fang_genos %>% 
    filter(str_detect(Group, "ZMPBA|ZMPIL|ZMPJA"))
maize_transposed <- rownames_to_column(as.data.frame(t(select(maize_genos, -(Sample_ID:Group)))), var = "SNP_ID")
teosinte_transposed <- rownames_to_column(as.data.frame(t(select(teosinte_genos, -(Sample_ID:Group)))), var = "SNP_ID")
maize_snps <- merge(snps, maize_transposed)
teosinte_snps <- merge(snps, teosinte_transposed)
maize_snps <- maize_snps %>% filter(!(Chromosome %in% c("multiple", "unknown") | Position %in% c("multiple", "unknown")))
teosinte_snps <- teosinte_snps %>% filter(!(Chromosome %in% c("multiple", "unknown") | Position %in% c("multiple", "unknown")))
maize_snps <- arrange(maize_snps, as.numeric(Chromosome), as.numeric(Position))
teosinte_snps <- arrange(teosinte_snps, as.numeric(Chromosome), as.numeric(Position))
maize_snps_desc <- arrange(maize_snps, as.numeric(Chromosome), desc(as.numeric(Position)))
teosinte_snps_desc <- arrange(teosinte_snps, as.numeric(Chromosome), desc(as.numeric(Position)))
maize_snps_desc <- maize_snps_desc %>% mutate(across(everything(), ~ gsub("\\?/\\?", "-/-", .)))
teosinte_snps_desc <- teosinte_snps_desc %>% mutate(across(everything(), ~ gsub("\\?/\\?", "-/-", .)))
```

