---
title: "R_Assignment_WAB"
output:
  html_document: default
editor_options: 
  chunk_output_type: console
---

working order

-   visualize
-   data inspection

for visualization, have one file with teosinte/maize snps, including group info? filter as before for maize/teo groups into one file, transpose (-(Sample_ID:JG_OTU)), merge, filter unknown/multiple - can visualize by group - make sure group column named as group if making one large file

need to label axes for graphs - what to use for y-axis on SNP distribution? Set to chromosome value but normalize all to 1 for equal height?

### R Assignment

# Initializing Tidyverse, loading source data files

```{r}
rm(list = ls())
library(tidyverse)
fang_genos <- read_tsv("Data/fang_et_al_genotypes.txt")
snp_pos <- read_tsv("Data/snp_position.txt")
```

# Data Inspection

We can see what the start of each file looks like, and see total dimensions.

```{r}
head(fang_genos)
dim(fang_genos)
head(snp_pos)
dim(snp_pos)
```

# Data Processing

Selects desired columns from SNP information - SNP_ID, Chromosome, Position Filters maize groups to separate data frame - ZMMIL, ZMMLR, ZMMMR Filters teosinte groups to separate data frame - ZMPBA, ZMPIL, ZMPJA

```{r}
snps <- select(snp_pos, SNP_ID, Chromosome, Position)
maize_genos <- fang_genos %>% 
    filter(str_detect(Group, "ZMMIL|ZMMLR|ZMMMR"))
teosinte_genos <- fang_genos %>% 
    filter(str_detect(Group, "ZMPBA|ZMPIL|ZMPJA"))
```

Selects needed columns from maize/teosinte genotype data frames, transposes and enforces as a data frame, and converts row names to column, given column title "SNP_ID" (SNP IDs were column titles in original data set so transposing turned them into row titles, making the rownames a column allows us to use it as the common column for joining to SNP position data)

```{r}
maize_transposed <- rownames_to_column(as.data.frame(t(select(maize_genos, -(Sample_ID:Group)))), var = "SNP_ID")
teosinte_transposed <- rownames_to_column(as.data.frame(t(select(teosinte_genos, -(Sample_ID:Group)))), var = "SNP_ID")
```

Merges transposed genotype data to SNP position data for both maize, teosinte. Common column used to join is SNP_ID

```{r}
maize_snps <- merge(snps, maize_transposed)
teosinte_snps <- merge(snps, teosinte_transposed)
```

Filters out SNPs with unknown or multiple positions - result files contain only SNPs with known position on a known chromosome

```{r}
maize_snps <- maize_snps %>% filter(!(Chromosome %in% c("multiple", "unknown") | Position %in% c("multiple", "unknown")))
teosinte_snps <- teosinte_snps %>% filter(!(Chromosome %in% c("multiple", "unknown") | Position %in% c("multiple", "unknown")))
```

Numerically sorts files by SNP position ascending, creates new set of data frames sorting files numerically by SNP position descending

```{r}
maize_snps <- arrange(maize_snps, as.numeric(Position))
teosinte_snps <- arrange(teosinte_snps, as.numeric(Position))
maize_snps_desc <- arrange(maize_snps, desc(as.numeric(Position)))
teosinte_snps_desc <- arrange(teosinte_snps, desc(as.numeric(Position)))
```

Replaces ?/? genotypes with -/- in descending sorted files

```{r}
maize_snps_desc[maize_snps_desc == "?/?"] <- "-/-"
teosinte_snps_desc[teosinte_snps_desc == "?/?"] <- "-/-"
```

Separates maize SNP data into data frames for each chromosome, both ascending and descending by position, saves as .RData files in /Result_Files folder

```{r}
maize_1 <- filter(maize_snps, Chromosome == 1)
save(maize_1, file="Result_Files/maize_1.RData")
maize_2 <- filter(maize_snps, Chromosome == 2)
save(maize_2, file="Result_Files/maize_2.RData")
maize_3 <- filter(maize_snps, Chromosome == 3)
save(maize_3, file="Result_Files/maize_3.RData")
maize_4 <- filter(maize_snps, Chromosome == 4)
save(maize_4, file="Result_Files/maize_4.RData")
maize_5 <- filter(maize_snps, Chromosome == 5)
save(maize_5, file="Result_Files/maize_5.RData")
maize_6 <- filter(maize_snps, Chromosome == 6)
save(maize_6, file="Result_Files/maize_6.RData")
maize_7 <- filter(maize_snps, Chromosome == 7)
save(maize_7, file="Result_Files/maize_7.RData")
maize_8 <- filter(maize_snps, Chromosome == 8)
save(maize_8, file="Result_Files/maize_8.RData")
maize_9 <- filter(maize_snps, Chromosome == 9)
save(maize_9, file="Result_Files/maize_9.RData")
maize_10 <- filter(maize_snps, Chromosome == 10)
save(maize_10, file="Result_Files/maize_10.RData")

maize_1_desc <- filter(maize_snps_desc, Chromosome == 1)
save(maize_1_desc, file="Result_Files/maize_1_desc.RData")
maize_2_desc <- filter(maize_snps_desc, Chromosome == 2)
save(maize_2_desc, file="Result_Files/maize_2_desc.RData")
maize_3_desc <- filter(maize_snps_desc, Chromosome == 3)
save(maize_3_desc, file="Result_Files/maize_3_desc.RData")
maize_4_desc <- filter(maize_snps_desc, Chromosome == 4)
save(maize_4_desc, file="Result_Files/maize_4_desc.RData")
maize_5_desc <- filter(maize_snps_desc, Chromosome == 5)
save(maize_5_desc, file="Result_Files/maize_5_desc.RData")
maize_6_desc <- filter(maize_snps_desc, Chromosome == 6)
save(maize_6_desc, file="Result_Files/maize_6_desc.RData")
maize_7_desc <- filter(maize_snps_desc, Chromosome == 7)
save(maize_7_desc, file="Result_Files/maize_7_desc.RData")
maize_8_desc <- filter(maize_snps_desc, Chromosome == 8)
save(maize_8_desc, file="Result_Files/maize_8_desc.RData")
maize_9_desc <- filter(maize_snps_desc, Chromosome == 9)
save(maize_9_desc, file="Result_Files/maize_9_desc.RData")
maize_10_desc <- filter(maize_snps_desc, Chromosome == 10)
save(maize_10_desc, file="Result_Files/maize_10_desc.RData")
```

Separates teosinte SNP data into data frames for each chromosome, both ascending and descending by position, saves as .RData files in /Result_Files folder

```{r}
teosinte_1 <- filter(teosinte_snps, Chromosome == 1)
save(teosinte_1, file="Result_Files/teosinte_1.RData")
teosinte_2 <- filter(teosinte_snps, Chromosome == 2)
save(teosinte_2, file="Result_Files/teosinte_2.RData")
teosinte_3 <- filter(teosinte_snps, Chromosome == 3)
save(teosinte_3, file="Result_Files/teosinte_3.RData")
teosinte_4 <- filter(teosinte_snps, Chromosome == 4)
save(teosinte_4, file="Result_Files/teosinte_4.RData")
teosinte_5 <- filter(teosinte_snps, Chromosome == 5)
save(teosinte_5, file="Result_Files/teosinte_5.RData")
teosinte_6 <- filter(teosinte_snps, Chromosome == 6)
save(teosinte_6, file="Result_Files/teosinte_6.RData")
teosinte_7 <- filter(teosinte_snps, Chromosome == 7)
save(teosinte_7, file="Result_Files/teosinte_7.RData")
teosinte_8 <- filter(teosinte_snps, Chromosome == 8)
save(teosinte_8, file="Result_Files/teosinte_8.RData")
teosinte_9 <- filter(teosinte_snps, Chromosome == 9)
save(teosinte_9, file="Result_Files/teosinte_9.RData")
teosinte_10 <- filter(teosinte_snps, Chromosome == 10)
save(teosinte_10, file="Result_Files/teosinte_10.RData")

teosinte_1_desc <- filter(teosinte_snps_desc, Chromosome == 1)
save(teosinte_1_desc, file="Result_Files/teosinte_1_desc.RData")
teosinte_2_desc <- filter(teosinte_snps_desc, Chromosome == 2)
save(teosinte_2_desc, file="Result_Files/teosinte_2_desc.RData")
teosinte_3_desc <- filter(teosinte_snps_desc, Chromosome == 3)
save(teosinte_3_desc, file="Result_Files/teosinte_3_desc.RData")
teosinte_4_desc <- filter(teosinte_snps_desc, Chromosome == 4)
save(teosinte_4_desc, file="Result_Files/teosinte_4_desc.RData")
teosinte_5_desc <- filter(teosinte_snps_desc, Chromosome == 5)
save(teosinte_5_desc, file="Result_Files/teosinte_5_desc.RData")
teosinte_6_desc <- filter(teosinte_snps_desc, Chromosome == 6)
save(teosinte_6_desc, file="Result_Files/teosinte_6_desc.RData")
teosinte_7_desc <- filter(teosinte_snps_desc, Chromosome == 7)
save(teosinte_7_desc, file="Result_Files/teosinte_7_desc.RData")
teosinte_8_desc <- filter(teosinte_snps_desc, Chromosome == 8)
save(teosinte_8_desc, file="Result_Files/teosinte_8_desc.RData")
teosinte_9_desc <- filter(teosinte_snps_desc, Chromosome == 9)
save(teosinte_9_desc, file="Result_Files/teosinte_9_desc.RData")
teosinte_10_desc <- filter(teosinte_snps_desc, Chromosome == 10)
save(teosinte_10_desc, file="Result_Files/teosinte_10_desc.RData")
```

# Data Visualization

Plotting position of SNPs on X axis from pter to qter for both maize and teosinte. No difference can be seen across each chromosome in each species - no difference in SNP distribution between maize and teosinte.

```{r}
ggplot(data=maize_snps, mapping=aes(x=Position, y=as.numeric(Chromosome))) + geom_point() + facet_wrap(~as.numeric(Chromosome), strip.position="bottom", scales="free_y") + theme(axis.text.y = element_blank(), axis.ticks.y = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank()) +ylab("Maize Chromosome")
ggplot(data=teosinte_snps, mapping=aes(x=Position, y=as.numeric(Chromosome))) + geom_point() + facet_wrap(~as.numeric(Chromosome), strip.position="bottom", scales="free_y") + theme(axis.text.y = element_blank(), axis.ticks.y = element_blank(), axis.text.x = element_blank(), axis.ticks.x = element_blank()) +ylab("Teosinte Chromosome")
```

Reformats SNP data files for maize, teosinte - pivots genotype data from columns to rows, assigns each to sample. Filters by genotype to denote zygosity, adds column. Adds column to indicate group - allows for comparisons between groups, samples.

```{r}
maize_vis <- pivot_longer(maize_snps, cols = !1:3, names_to = "Sample", values_to = "Genotype")
maize_vis$Zygosity <- case_when(maize_vis$Genotype == "A/A" ~ "homozygous", maize_vis$Genotype == "C/C" ~ "homozygous", maize_vis$Genotype == "G/G" ~ "homozygous", maize_vis$Genotype == "T/T" ~ "homozygous", maize_vis$Genotype == "?/?" ~ "missing", .default = "heterozygous" )
teosinte_vis <- pivot_longer(teosinte_snps, cols = !1:3, names_to = "Sample", values_to = "Genotype")
teosinte_vis$Zygosity <- case_when(teosinte_vis$Genotype == "A/A" ~ "homozygous", teosinte_vis$Genotype == "C/C" ~ "homozygous", teosinte_vis$Genotype == "G/G" ~ "homozygous", teosinte_vis$Genotype == "T/T" ~ "homozygous", teosinte_vis$Genotype == "?/?" ~ "missing", .default = "heterozygous" )
maize_vis$Group <- "Maize"
teosinte_vis$Group <- "Teosinte"
```

Combines maize and teosinte visualization data frames into one for comparisons between groups

```{r}
combined_vis <- rbind(maize_vis, teosinte_vis)
```

```{r}
combined_vis$zygosity_value <- ifelse(combined_vis$Zygosity == "homozygous", 1,
         ifelse(combined_vis$Zygosity == "heterozygous", 0, NA))
```



